pkg/KubeArmorOperator/internal/controller/cluster.go

```diff
@@ -0,0 +1,391 @@
+							clusterWatcher.Nodes[i].Enforcer != newNode.Enforcer ||
+							clusterWatcher.Nodes[i].Name != newNode.Name ||
+							clusterWatcher.Nodes[i].Runtime != newNode.Runtime ||
+							clusterWatcher.Nodes[i].RuntimeSocket != newNode.RuntimeSocket ||
+							clusterWatcher.Nodes[i].RuntimeStorage != newNode.RuntimeStorage {
+							clusterWatcher.Nodes[i] = newNode
+							nodeModified = true
+							clusterWatcher.Log.Infof("Node %s was updated", node.Name)
+						}
+					}
+					clusterWatcher.NodesLock.Unlock()
+					if nodeModified {
+						clusterWatcher.UpdateDaemonsets(common.DeletAction, newNode.Enforcer, newNode.Runtime, newNode.RuntimeSocket, newNode.RuntimeStorage)
+					}
+					clusterWatcher.UpdateDaemonsets(common.AddAction, newNode.Enforcer, newNode.Runtime, newNode.RuntimeSocket, newNode.RuntimeStorage)
+				}
+			} else {
+				log.Errorf("Cannot convert object to node struct")
+				log.Error(newObj)
+			}
+		},
+		DeleteFunc: func(obj interface{}) {
+			if node, ok := obj.(*corev1.Node); ok {
+				deletedNode := Node{}
+				clusterWatcher.NodesLock.Lock()
+				for i, n := range clusterWatcher.Nodes {
+					if n.Name == node.Name {
+						clusterWatcher.Nodes = append(clusterWatcher.Nodes[:i], clusterWatcher.Nodes[i+1:]...)
+						deletedNode = n
+						break
+					}
+				}
+				clusterWatcher.NodesLock.Unlock()
+				clusterWatcher.UpdateDaemonsets(common.DeletAction, deletedNode.Enforcer, deletedNode.Runtime, deletedNode.RuntimeSocket, deletedNode.RuntimeStorage)
+			}
+		},
+	})
+
+	nodeInformer.Run(wait.NeverStop)
+}
+
+func (clusterWatcher *ClusterWatcher) UpdateDaemonsets(action, enforcer, runtime, socket, runtimeStorage string) {
+	clusterWatcher.Log.Info("updating daemonset")
+	daemonsetName := strings.Join([]string{
+		"kubearmor",
+		strings.ReplaceAll(enforcer, ".", "-"),
+		runtime,
+		common.ShortSHA(socket),
+	}, "-")
+	newDaemonSet := false
+	deleteDaemonSet := false
+	clusterWatcher.DaemonsetsLock.Lock()
+	if action == common.AddAction {
+		clusterWatcher.Daemonsets[daemonsetName]++
+		_, err := clusterWatcher.Client.AppsV1().DaemonSets(common.Namespace).Get(context.Background(), daemonsetName, v1.GetOptions{})
+		if err != nil {
+			newDaemonSet = true
+		}
+	} else if action == common.DeletAction {
+		if val, ok := clusterWatcher.Daemonsets[daemonsetName]; ok {
+			if val < 2 {
+				clusterWatcher.Daemonsets[daemonsetName] = 0
+				deleteDaemonSet = true
+			} else {
+				clusterWatcher.Daemonsets[daemonsetName]--
+			}
+		}
+	}
+	clusterWatcher.DaemonsetsLock.Unlock()
+
+	if deleteDaemonSet {
+		err := clusterWatcher.Client.AppsV1().DaemonSets(common.Namespace).Delete(context.Background(), daemonsetName, v1.DeleteOptions{})
+		if err != nil {
+			clusterWatcher.Log.Warnf("Cannot delete daemonset %s, error=%s", daemonsetName, err.Error())
+		}
+	}
+	if newDaemonSet {
+		daemonset := generateDaemonset(daemonsetName, enforcer, runtime, socket, runtimeStorage)
```

Comment:
Question: Currently we're only creating a single daemonset for all kinds of nodes, right? Are we handling if different nodes have different enforcers or container runtimes?

Reply:
https://github.com/kubearmor/KubeArmor/pull/1246#pullrequestreview-1513832099

@daemon1024 this might be what you were looking for?

Reply:
> Are we handling if different nodes have different enforcers or container runtimes?
yes we're handling this case, we are probing each node for container runtime and enforcer and deploying the daemonset accordingly.